module/zfs/metaslab.c

```diff
@@ -1760,10 +1769,18 @@ metaslab_df_alloc(metaslab_t *msp, uint64_t size)
 	 */
 	if (metaslab_largest_allocatable(msp) < metaslab_df_alloc_threshold ||
 	    free_pct < metaslab_df_free_pct) {
+		align = size & -size;
+		cursor = &msp->ms_lbas[highbit64(align) - 1];
 		offset = -1;
 	} else {
-		offset = metaslab_block_picker(rt,
-		    cursor, size, metaslab_df_max_search);
+		offset = metaslab_block_picker(rt, cursor, size, max_size,
+		    metaslab_df_max_search, found_size);
+		if (max_size != size && offset == -1) {
+			align = size & -size;
+			cursor = &msp->ms_lbas[highbit64(align) - 1];
+			offset = metaslab_block_picker(rt, cursor, max_size,
+			    max_size, metaslab_df_max_search, found_size);
+		}
```

Comment:
I don't understand this part.  First you are looking into the area of maximum-sized requests, but allowing for shorter one, which is OK.  But then you are looking in the area of shorter requests strictly for a large one, which makes no sense to me.  If we don't know what we'll get allocated, all this allocation optimization logic does not make sense.  I am generally not sure all this magic with areas makes much sense these days and does not just produce more fragmentation and head seeks, but this added code I just don't understand.